mpcs/src/util/merkle_tree.rs

```diff
@@ -23,36 +23,89 @@ use ark_std::{end_timer, start_timer};
 use super::hash::write_digest_to_transcript;
 
 #[derive(Clone, Debug, Default, Serialize, Deserialize)]
-#[serde(bound(serialize = "E: Serialize", deserialize = "E: DeserializeOwned"))]
-pub struct MerkleTree<E: ExtensionField>
+#[serde(bound(deserialize = "E: DeserializeOwned"))]
+pub struct MerkleTreeDigests<E: ExtensionField>
 where
     E::BaseField: Serialize + DeserializeOwned,
 {
     inner: Vec<Vec<Digest<E::BaseField>>>,
-    leaves: Vec<FieldType<E>>,
 }
 
-impl<E: ExtensionField> MerkleTree<E>
+impl<E: ExtensionField> MerkleTreeDigests<E>
 where
     E::BaseField: Serialize + DeserializeOwned,
 {
-    pub fn compute_inner(leaves: &FieldType<E>) -> Vec<Vec<Digest<E::BaseField>>> {
-        merkelize::<E>(&[leaves])
+    pub fn from_leaves(leaves: &FieldType<E>) -> Self {
```

Comment:
Consider implementing `From` instances instead?  (Not completely sure.)

Reply:
I'm not sure, either. Computing the Merkle tree is an expensive procedure, so we probably want it to be always invoked explicitly.

Reply:
Rust never invokes `from` implicitly.

Reply:
No, not in the language. I just have the impression (which may be wrong) that `.into()` gives people the hint that this is a cheap type conversion, invoking some wrapper functions, without any expensive stuff.